// sphere-node-sdk/lib/services/base.js
var BaseService, GraphQLError, HttpError, Promise, REGEX_LAST, SphereHttpError, Utils, _, debug, ref;

debug = require('debug')('sphere-client');

_ = require('underscore');

_.mixin(require('underscore-mixins'));

Promise = require('bluebird');

Utils = require('../utils');

ref = require('../errors'), HttpError = ref.HttpError, GraphQLError = ref.GraphQLError, SphereHttpError = ref.SphereHttpError;

REGEX_LAST = /^(\d+)([s|m|h|d|w])$/;

BaseService = (function() {
  BaseService.baseResourceEndpoint = '';

  BaseService.supportsByKey = false;

  function BaseService(options) {
    if (options == null) {
      options = {};
    }
    this._rest = options._rest, this._task = options._task, this._stats = options._stats;
    this._setDefaults();
  }

  BaseService.prototype._setDefaults = function() {
    this._currentEndpoint = this.constructor.baseResourceEndpoint;
    this._params = {
      encoded: ['where', 'expand', 'sort'],
      plain: ['perPage', 'page'],
      query: {
        where: [],
        operator: 'and',
        sort: [],
        expand: []
      }
    };
    return this._fetchAll = false;
  };

  BaseService.prototype.byId = function(id) {
    if (this._params.key) {
      throw new Error("You cannot fetch or update by Id and Key at the same time");
    }
    this._currentEndpoint = this.constructor.baseResourceEndpoint + "/" + id;
    this._params.id = id;
    debug('setting endpoint id: %j', this._currentEndpoint);
    return this;
  };

  BaseService.prototype.byKey = function(key) {
    if (this._params.id) {
      throw new Error("You cannot fetch or update by Id and Key at the same time");
    }
    this._currentEndpoint = this.constructor.baseResourceEndpoint + "/key=" + key;
    this._params.key = key;
    debug('setting endpoint key: %j', this._currentEndpoint);
    return this;
  };

  BaseService.prototype.where = function(predicate) {
    var encodedPredicate;
    if (!predicate) {
      return this;
    }
    encodedPredicate = encodeURIComponent(predicate);
    this._params.query.where.push(encodedPredicate);
    debug('setting predicate: %s', predicate);
    return this;
  };

  BaseService.prototype.whereOperator = function(operator) {
    if (operator == null) {
      operator = 'and';
    }
    this._params.query.operator = (function() {
      switch (operator) {
        case 'and':
        case 'or':
          return operator;
        default:
          return 'and';
      }
    })();
    debug('setting where operator: %s', operator);
    return this;
  };

  BaseService.prototype.last = function(period) {
    var amount, before, dateTime, matches, now;
    if (!REGEX_LAST.test(period)) {
      throw new Error("Cannot parse period '" + period + "'");
    }
    matches = REGEX_LAST.exec(period);
    amount = matches[1];
    if (amount === '0') {
      return this;
    }
    before = Utils.getTime(amount, matches[2]);
    now = new Date().getTime();
    dateTime = new Date(now - before).toISOString();
    return this.where("lastModifiedAt > \"" + dateTime + "\"");
  };

  BaseService.prototype.sort = function(path, ascending) {
    var direction;
    if (ascending == null) {
      ascending = true;
    }
    direction = ascending ? 'asc' : 'desc';
    this._params.query.sort.push(encodeURIComponent(path + " " + direction));
    debug('setting sort: %s %s', path, direction);
    return this;
  };

  BaseService.prototype.page = function(page) {
    if (_.isNumber(page) && page < 1) {
      throw new Error('Page must be a number >= 1');
    }
    this._params.query.page = page;
    debug('setting page: %s', page);
    return this;
  };

  BaseService.prototype.perPage = function(perPage) {
    if (_.isNumber(perPage) && perPage < 0) {
      throw new Error('PerPage (limit) must be a number >= 0');
    }
    this._params.query.perPage = perPage;
    debug('setting perPage: %s', perPage);
    return this;
  };

  BaseService.prototype.all = function() {
    this._fetchAll = true;
    return this;
  };

  BaseService.prototype.expand = function(expansionPath) {
    var encodedExpansionPath;
    if (!expansionPath) {
      return this;
    }
    encodedExpansionPath = encodeURIComponent(expansionPath);
    this._params.query.expand.push(encodedExpansionPath);
    debug('setting expand: %s', expansionPath);
    return this;
  };

  BaseService.prototype.byQueryString = function(query, withEncodedParams) {
    var parsed;
    if (withEncodedParams == null) {
      withEncodedParams = false;
    }
    parsed = _.parseQuery(query, false);
    if (!withEncodedParams) {
      _.each(this._params.encoded, (function(_this) {
        return function(param) {
          if (parsed[param]) {
            if (_.contains(_this._params.encoded, param)) {
              parsed[param] = _.map(_.flatten([parsed[param]]), function(p) {
                return encodeURIComponent(p);
              });
              return _this._params.query[param] = parsed[param];
            }
          }
        };
      })(this));
      _.each(this._params.plain, (function(_this) {
        return function(param) {
          if (parsed[param]) {
            return _this._params.query[param] = parsed[param];
          }
        };
      })(this));
    }
    this._params.queryString = _.stringifyQuery(parsed);
    debug('setting queryString: %s', query);
    return this;
  };

  BaseService.prototype._queryString = function() {
    var qs;
    qs = this._params.queryString ? this._params.queryString : Utils.buildQueryString({
      where: this._params.query.where,
      whereOperator: this._params.query.operator,
      page: this._params.query.page,
      perPage: this._params.query.perPage,
      sort: this._params.query.sort,
      expand: this._params.query.expand
    });
    debug('built query string: %s', qs);
    return qs;
  };

  BaseService.prototype.fetch = function() {
    var _getEndpoint;
    _getEndpoint = (function(_this) {
      return function() {
        var endpoint, queryString;
        queryString = _this._queryString();
        endpoint = _this._currentEndpoint;
        if (queryString) {
          endpoint += "?" + queryString;
        }
        return endpoint;
      };
    })(this);
    if (!this._params.queryString && this._fetchAll === true) {
      if (_.isEmpty(this._params.query.sort)) {
        this.sort('id');
      }
      return this._paged(_getEndpoint());
    } else {
      return this._get(_getEndpoint());
    }
  };

  BaseService.prototype.process = function(fn, options) {
    if (options == null) {
      options = {};
    }
    if (!_.isFunction(fn)) {
      throw new Error('Please provide a function to process the elements');
    }
    return new Promise((function(_this) {
      return function(resolve, reject) {
        var _processPage, endpoint, originalPredicate, originalQuery;
        options = _.defaults(options, {
          accumulate: true
        });
        endpoint = _this.constructor.baseResourceEndpoint;
        originalQuery = _this._params.query;
        originalPredicate = _this._params.query.where;
        if (originalQuery.perPage === 0) {
          debug('using batch size of 20 since 0 wont return any results');
          originalQuery.perPage = 20;
        }
        _processPage = function(lastId, acc) {
          var enhancedQueryString, lastIdPredicate, queryString, wherePredicate;
          if (acc == null) {
            acc = [];
          }
          debug('processing next page with id: %s', lastId);
          if (_.isEmpty(_this._params.query.sort)) {
            _this.sort('id');
          }
          _this._params.query = _.extend({}, originalQuery, {
            where: []
          });
          queryString = _this._queryString();
          wherePredicate = originalPredicate.join(encodeURIComponent(" " + (originalQuery.operator || 'and') + " "));
          if (lastId) {
            lastIdPredicate = encodeURIComponent("id > \"" + lastId + "\"");
            wherePredicate = _.isEmpty(originalPredicate) ? lastIdPredicate : wherePredicate + "%20and%20" + lastIdPredicate;
          }
          debug('process where predicate: %j', wherePredicate);
          enhancedQueryString = [queryString, "withTotal=false"].join('&');
          if (wherePredicate) {
            enhancedQueryString = enhancedQueryString + "&where=" + wherePredicate;
          }
          debug('enhanced query: %s', enhancedQueryString);
          return _this._get(endpoint + "?" + enhancedQueryString).then(function(payload) {
            return fn(payload).then(function(result) {
              var accumulated, last, newLastId, resultsSize;
              resultsSize = _.size(payload.body.results);
              debug('accumulating: size %s, offset %s, count %s', resultsSize, payload.body.offset, payload.body.count);
              if (options.accumulate) {
                accumulated = acc.concat(result || []);
              }
              if (resultsSize < (originalQuery.perPage || 20)) {
                return resolve(accumulated || []);
              }
              last = _.last(payload.body.results);
              newLastId = last && last.id;
              return _processPage(newLastId, accumulated);
            });
          })["catch"](function(error) {
            return reject(error);
          }).done();
        };
        return _processPage();
      };
    })(this));
  };

  BaseService.prototype.save = function(body) {
    var endpoint, queryString;
    if (!body) {
      throw new Error("Body payload is required for creating a resource (endpoint: " + this.constructor.baseResourceEndpoint + ")");
    }
    queryString = Utils.buildQueryString({
      expand: this._params.query.expand
    });
    endpoint = this.constructor.baseResourceEndpoint;
    if (queryString) {
      endpoint += "?" + queryString;
    }
    return this._save(endpoint, body);
  };

  BaseService.prototype.create = function() {
    return this.save.apply(this, arguments);
  };

  BaseService.prototype.update = function(body) {
    var endpoint, queryString;
    if (this.constructor.supportsByKey === true) {
      if (!(this._params.key || this._params.id)) {
        if (!(this._params.key || this._params.id)) {
          throw new Error("Missing resource id. You can set it by chaining '.byId(ID)' or '.byKey(KEY)'");
        }
      }
    } else {
      if (!this._params.id) {
        throw new Error("Missing resource id. You can set it by chaining '.byId(ID)'");
      }
    }
    if (!body) {
      throw new Error("Body payload is required for updating a resource (endpoint: " + this._currentEndpoint + ")");
    }
    queryString = Utils.buildQueryString({
      expand: this._params.query.expand
    });
    endpoint = this._currentEndpoint;
    if (queryString) {
      endpoint += "?" + queryString;
    }
    return this._save(endpoint, body);
  };

  BaseService.prototype["delete"] = function(version) {
    var endpoint;
    if (!version) {
      throw new Error("Version is required for deleting a resource (endpoint: " + this._currentEndpoint + ")");
    }
    if (this.constructor.supportsByKey === true) {
      if (!(this._params.key || this._params.id)) {
        if (!(this._params.key || this._params.id)) {
          throw new Error("Missing resource id. You can set it by chaining '.byId(ID)' or '.byKey(KEY)'");
        }
      }
    } else {
      if (!this._params.id) {
        throw new Error("Missing resource id. You can set it by chaining '.byId(ID)'");
      }
    }
    endpoint = this._currentEndpoint + "?version=" + version;
    return this._delete(endpoint);
  };

  BaseService.prototype._get = function(endpoint) {
    this._setDefaults();
    return this._task.addTask((function(_this) {
      return function() {
        var originalRequest;
        originalRequest = {
          endpoint: endpoint
        };
        return new Promise(function(resolve, reject) {
          return _this._rest.GET(endpoint, function() {
            return _this._wrapResponse.apply(_this, [resolve, reject, originalRequest].concat(_.toArray(arguments)));
          });
        });
      };
    })(this));
  };

  BaseService.prototype._paged = function(endpoint) {
    this._setDefaults();
    return this._task.addTask((function(_this) {
      return function() {
        var originalRequest;
        originalRequest = {
          endpoint: endpoint
        };
        return new Promise(function(resolve, reject) {
          return _this._rest.PAGED(endpoint, function() {
            return _this._wrapResponse.apply(_this, [resolve, reject, originalRequest].concat(_.toArray(arguments)));
          });
        });
      };
    })(this));
  };

  BaseService.prototype._save = function(endpoint, payload) {
    this._setDefaults();
    return this._task.addTask((function(_this) {
      return function() {
        var originalRequest;
        originalRequest = {
          endpoint: endpoint,
          payload: payload
        };
        return new Promise(function(resolve, reject) {
          return _this._rest.POST(endpoint, payload, function() {
            return _this._wrapResponse.apply(_this, [resolve, reject, originalRequest].concat(_.toArray(arguments)));
          });
        });
      };
    })(this));
  };

  BaseService.prototype._delete = function(endpoint) {
    this._setDefaults();
    return this._task.addTask((function(_this) {
      return function() {
        var originalRequest;
        originalRequest = {
          endpoint: endpoint
        };
        return new Promise(function(resolve, reject) {
          return _this._rest.DELETE(endpoint, function() {
            return _this._wrapResponse.apply(_this, [resolve, reject, originalRequest].concat(_.toArray(arguments)));
          });
        });
      };
    })(this));
  };

  BaseService.prototype._wrapResponse = function(resolve, reject, originalRequest, error, response, body) {
    var endpoint, errMsg, errorBody, errorMessage, errorResp, graphqlError, options, ref1, ref2, responseJson;
    if (this._stats.maskSensitiveHeaderData === true && response) {
      response.req._header = Utils._censorHeaderStr(response.req._header);
      response.request.headers = Utils._censorHeaderObj(response.request.headers);
      response.headers = Utils._censorHeaderObj(response.headers);
    }
    responseJson = this._stats.includeHeaders && response ? {
      http: {
        request: {
          method: response.request.method,
          httpVersion: response.httpVersion,
          uri: response.request.uri,
          header: response.req._header,
          headers: response.request.headers
        },
        response: {
          headers: response.headers
        }
      }
    } : {};
    options = _.deepClone(_.omit(this._rest._options, ['access_token']));
    if (options.config) {
      options.config = _.omit(options.config, ['client_id', 'client_secret']);
    }
    if (this._stats.maskSensitiveHeaderData === true) {
      options.headers = Utils._censorHeaderObj(options.headers);
    }
    originalRequest.options = options;
    if (error) {
      if (error instanceof Error) {
        errorMessage = error.message;
      } else {
        errorMessage = error;
      }
      errorResp = {
        statusCode: response != null ? response.statusCode : void 0,
        message: errorMessage,
        originalRequest: originalRequest
      };
      if (body) {
        errorResp.body = body;
      }
      errorBody = _.extend(responseJson, errorResp);
      return reject(new HttpError(errorMessage, errorBody));
    } else {
      if ((ref1 = response.headers) != null ? ref1['X-DEPRECATION-NOTICE'] : void 0) {
        console.warn("Deprecation notice: " + response.headers['X-DEPRECATION-NOTICE']);
      }
      if ((200 <= (ref2 = response.statusCode) && ref2 < 300)) {
        return resolve(_.extend(responseJson, {
          statusCode: response.statusCode,
          body: body
        }));
      } else if (response.statusCode === 404) {
        endpoint = response.request.uri.path;
        return reject(new SphereHttpError.NotFound("Endpoint '" + endpoint + "' not found.", _.extend(responseJson, {
          statusCode: response.statusCode,
          message: "Endpoint '" + endpoint + "' not found.",
          originalRequest: originalRequest
        })));
      } else if (response.statusCode === 405) {
        endpoint = response.request.uri.path;
        errMsg = "The chosen HTTP method is not allowed for the endpoint '" + endpoint + "'. This might be caused by a malformed request.";
        return reject(new SphereHttpError.MethodNotAllowed(errMsg, _.extend(responseJson, {
          statusCode: response.statusCode,
          message: errMsg,
          originalRequest: originalRequest
        })));
      } else if (response.statusCode === 400 && this.constructor.name === 'GraphQLService' && !body.data) {
        graphqlError = new GraphQLError('GraphQL error', _.extend(responseJson, body, {
          statusCode: 400,
          originalRequest: originalRequest
        }));
        return reject(graphqlError);
      } else {
        errorMessage = body.message || body.error_description || body.error || 'Undefined SPHERE.IO error message';
        errorBody = _.extend(responseJson, body, {
          statusCode: body.statusCode || response.statusCode,
          originalRequest: originalRequest
        });
        return reject((function() {
          switch (errorBody.statusCode) {
            case 400:
              return new SphereHttpError.BadRequest(errorMessage, errorBody);
            case 401:
              return new SphereHttpError.Unauthorized(errorMessage, errorBody);
            case 409:
              return new SphereHttpError.ConcurrentModification(errorMessage, errorBody);
            case 500:
              return new SphereHttpError.InternalServerError(errorMessage, errorBody);
            case 503:
              return new SphereHttpError.ServiceUnavailable(errorMessage, errorBody);
            default:
              return new HttpError(require('http').STATUS_CODES[response.statusCode], _.extend(responseJson, {
                statusCode: response.statusCode,
                originalRequest: originalRequest
              }));
          }
        })());
      }
    }
  };

  return BaseService;

})();

module.exports = BaseService;
