// sphere-node-utils/lib/helpers/extended-logger.js
var Logger, _;

_ = require('underscore');

Logger = require('./logger');


/**
 * Creates a new instance of the ExtendedLogger
 * @class ExtendedLogger
 */

module.exports = (function() {
  function _Class(options) {
    var logConfig, ref;
    if (options == null) {
      options = {};
    }
    ref = _.defaults(options, {
      additionalFields: {}
    }), logConfig = ref.logConfig, this.additionalFields = ref.additionalFields;
    this.tmpAdditionalFields = {};
    this.bunyanLogger = new Logger(logConfig);
  }

  _Class.prototype._serializeError = function(e) {
    return {
      body: e.body,
      message: e.message,
      name: e.name,
      stack: e.stack,
      code: e.code,
      statusCode: e.statusCode,
      signal: e.signal
    };
  };

  _Class.prototype._wrapOptions = function(type, opts, msg) {
    var wrappedData;
    if (!msg && _.isString(opts)) {
      msg = opts;
      opts = {};
    }
    if (opts instanceof Error) {
      opts = this._serializeError(opts);
    } else if ((opts != null ? opts.err : void 0) instanceof Error) {
      opts.err = this._serializeError(opts.err);
    }
    wrappedData = {
      data: opts
    };
    _.extend(wrappedData, this.additionalFields, this.tmpAdditionalFields);
    this.tmpAdditionalFields = {};
    return this.bunyanLogger[type](wrappedData, msg);
  };

  _Class.prototype.withField = function(obj) {
    _.extend(this.tmpAdditionalFields, obj);
    return this;
  };

  _Class.prototype.trace = function(opts, msg) {
    return this._wrapOptions('trace', opts, msg);
  };

  _Class.prototype.debug = function(opts, msg) {
    return this._wrapOptions('debug', opts, msg);
  };

  _Class.prototype.info = function(opts, msg) {
    return this._wrapOptions('info', opts, msg);
  };

  _Class.prototype.warn = function(opts, msg) {
    return this._wrapOptions('warn', opts, msg);
  };

  _Class.prototype.error = function(opts, msg) {
    return this._wrapOptions('error', opts, msg);
  };

  _Class.prototype.fatal = function(opts, msg) {
    return this._wrapOptions('fatal', opts, msg);
  };

  return _Class;

})();
